### PR TITLE
[RyuJIT/ARM32] Fix resolving TYP_DOUBLE and TYP_FLOAT intervals

### PR DESCRIPTION
**Pull request by hqueue**
When generating fromReg to toReg move to resolve edges,
we have to consider either of fromReg and toReg can contain double type.

If TYP_DOUBLE interval occupies fromReg/toReg, we have to look up two float registers when making decisions.

Three are various cases as below and case 4 causes an assertion from  `JIT\Regression\CLR-x86-JIT\V1-M12-Beta2\b59782\b59782\b59782.cmd`  at issue #14374.

1. From dX (double interval) to dY (double interval) // No problem
2. From dX (double interval) to **dY (two float intervals)** // Fixed by this PR
3. From dX (double interval) to dY (1st float reg is assigned to a float interval) // No problem
4. From dX (double interval) to **dY (2nd float reg is assigned to a float interval)** // Fixed by this PR
5. From dX (double interval) to dY (free) // No problem

6. From sX (float interval) to sY (float interval) // No problem 
7. From sX (float interval) to sY (part of a double interval) // I'm still looking at this case.
8. From sX (float interval) to sY (free) // No problem

This PR partially resolve an assertion in #14374.
`JIT\Regression\CLR-x86-JIT\V1-M12-Beta2\b59782\b59782\b59782.cmd` is passed with this PR.

And remaining assertion is being previously addressed by PR #14588.

### CODE REVIEW

**Comment by hqueue** [comment]
@CarolEidt This is another PR to fix a remaining bug in resolving edges.
However I'm not sure about my fix. Please take a look and give any opinion.

File: src/jit/lsra.cpp [hunk 1]
```diff
@@ -10783,7 +10783,23 @@ void LinearScan::resolveEdge(BasicBlock*      fromBlock,
...
+                assert(genIsValidDoubleReg(targetReg));
+                regNumber anotherHalfRegNum = REG_NEXT(targetReg);
+                if (location[anotherHalfRegNum] == REG_NA)
+                {
+                    targetRegsReady |= targetRegMask;
+                }
+            }
+            else
+#endif // _TARGET_ARM_
+                targetRegsReady |= targetRegMask;
```
Comment (hqueue) [inline src/jit/lsra.cpp]: Is this right place to add logic to handle TYP_DOUBLE ?
Because I assumes that `targetRegsReady` contains regMask of free registers, I added code to consider two float registers here.

File: src/jit/lsra.cpp [hunk 2]
```diff
@@ -10929,11 +10945,57 @@ void LinearScan::resolveEdge(BasicBlock*      fromBlock,
...
                 {
                     compiler->codeGen->regSet.rsSetRegsModified(genRegMask(tempReg) DEBUGARG(dumpTerse));
-                    assert(sourceIntervals[targetReg] != nullptr);
-                    addResolution(block, insertionPoint, sourceIntervals[targetReg], tempReg, targetReg);
-                    JITDUMP(" (%s)\n", resolveTypeName[resolveType]);
-                    location[targetReg] = (regNumberSmall)tempReg;
-                    targetRegsReady |= targetRegMask;
+#ifdef _TARGET_ARM_
+                    if (sourceIntervals[fromReg]->registerType == TYP_DOUBLE)
+                    {
```
Comment (hqueue) [inline src/jit/lsra.cpp]: I'm going to add a helper to encapsulate this part, such as `addResolutionForDouble`, if this logic is correct.

**Comment by hqueue** [comment]
@dotnet-bot please test Tizen armel Cross Debug Build

File: src/jit/lsra.cpp [hunk 3]
```diff
@@ -10929,11 +10945,58 @@ void LinearScan::resolveEdge(BasicBlock*      fromBlock,
                 else
                 {
                     compiler->codeGen->regSet.rsSetRegsModified(genRegMask(tempReg) DEBUGARG(dumpTerse));
-                    assert(sourceIntervals[targetReg] != nullptr);
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: This should be moved to the else clause instead of deleted.

File: src/jit/lsra.cpp [hunk 4]
```diff
@@ -10783,7 +10783,23 @@ void LinearScan::resolveEdge(BasicBlock*      fromBlock,
...
         if (location[targetReg] == REG_NA)
         {
-            targetRegsReady |= targetRegMask;
+#ifdef _TARGET_ARM_
+            regNumber sourceReg = (regNumber)source[targetReg];
+            Interval* interval  = sourceIntervals[sourceReg];
+            if (interval->registerType == TYP_DOUBLE)
+            {
+                // For ARM32, make sure that two float reigster
+                // consisting a double register are not used.
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: Please reword this as:
```
// For ARM32, make sure that both of the float halves of the double register are available.
```

**Comment by hqueue** [comment]
@dotnet-bot test ci help

**Comment by hqueue** [comment]
@dotnet-bot help

File: src/jit/lsra.cpp (hunk 1 above)
Comment (hqueue) [reply src/jit/lsra.cpp]: I see. Thanks.

File: src/jit/lsra.cpp (hunk 2 above)
Comment (hqueue) [reply src/jit/lsra.cpp]: Thanks. I will encapsulate this.

File: src/jit/lsra.cpp (hunk 4 above)
Comment (hqueue) [reply src/jit/lsra.cpp]: Got it :)

File: src/jit/lsra.cpp (hunk 3 above)
Comment (hqueue) [reply src/jit/lsra.cpp]: Right. Thanks.

**Comment by hqueue** [comment]
> I think there's another problem, though, that's not directly related to this change. I don't think that the code in resolveEdge() that gets a temp reg (https://github.com/dotnet/coreclr/blob/master/src/jit/lsra.cpp#L10677) is quite correct. If it can't find a double reg, it could find a float reg that is, in fact, a valid double register, even though its other half is not free. I think that we need a tempRegDbl for ARM. If we get a valid double reg, we set it, as well as tempRegFlt, to that register. Otherwise, we leave tempRegDbl as REG_NA.

Or I think we can just update https://github.com/dotnet/coreclr/blob/master/src/jit/lsra.cpp#L10844 to check validness of tempRegFlt for TYP_DOUBLE.
What do you think of it ?

**Comment by CarolEidt** [comment]
> I think we can just update https://github.com/dotnet/coreclr/blob/master/src/jit/lsra.cpp#L10844 to check validness of tempRegFlt for TYP_DOUBLE.
What do you think of it ?

I don't think that would be easy, because we'd in any case have to keep a separate flag or something to indicate whether both halves are available.

**Comment by hqueue** [comment]
@CarolEidt I've update the PR with your feedback.
And also introduce `tempRegDbl`. I came to same conclusion as your thought that using `temprRegDbl` results in simple and intuitive code after implementing several approaches. Thanks.!

File: src/jit/lsra.cpp [hunk 5]
```diff
@@ -10676,20 +10743,23 @@ void LinearScan::resolveEdge(BasicBlock*      fromBlock,
         (resolveType == ResolveSharedCritical) ? REG_NA : getTempRegForResolution(fromBlock, toBlock, TYP_INT);
 #endif // !_TARGET_XARCH_
     regNumber tempRegFlt = REG_NA;
+#ifdef _TARGET_ARM_
```
Comment (CarolEidt) [inline src/jit/lsra.cpp]: I would suggest that you not ifdef this out, if we can get away with it without the compiler(s) complaining. It will be optimized away in any case. I find this:
```
regNumber tempRegDbl = REG_NA; // Used only for ARM
```
easier on the eyes than the `#ifdef` and `#endif`

Comment (hqueue) [reply src/jit/lsra.cpp]: Sure :)

**Comment by hqueue** [comment]
I've squashed commits and remove one `#ifdef _TARGET_ARM_` as suggested.
